lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -445,9 +445,9 @@ static SILBasicBlock::iterator
 eliminateSimpleBorrows(BeginBorrowInst *bbi, CanonicalizeInstruction &pass) {
   auto next = std::next(bbi->getIterator());
 
-  // Never eliminate lexical borrow scopes.  They must be kept to ensure that
-  // value lifetimes aren't observably shortened.
-  if (bbi->isLexical())
+  // Never eliminate non-redundant lexical borrow scopes.  They must be kept to
+  // ensure that value lifetimes aren't observably shortened.
+  if (bbi->isLexical() && !isNestedLexicalBeginBorrow(bbi))
```

Comment:
You can't do this if the borrow is needed for diagnostics. For now, I suppose you can just check that the SIL stage is canonical.

For example, we already do this in splitAggregateLoad.

Reply:
Fixed.  Lexical + raw stage -> no elimination.